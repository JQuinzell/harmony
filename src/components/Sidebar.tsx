import { Grid, makeStyles } from '@material-ui/core'
import { observer } from 'mobx-react-lite'
import React from 'react'
import { useRootStore } from '../RootStoreContext'
import { AddServer } from './AddServer'
import { SearchServer } from './SearchServer'
import { ServerButton } from './ServerButton'

const useStyles = makeStyles((theme) => ({
  nav: {
    background: theme.palette.background.paper,
    height: '100%',
  },
  bottomIcons: {
    marginTop: 'auto',
  },
}))

export const SideBar: React.FC = observer(() => {
  const styles = useStyles()
  const rootStore = useRootStore()

  return (
    <Grid className={styles.nav} container direction='column' component='nav'>
      {rootStore.joinedServers.map(({ title }) => (
        <Grid key={title} item>
          <ServerButton onClick={() => rootStore.selectServer(title)}>
            {title.slice(0, 2)}
          </ServerButton>
        </Grid>
      ))}
      <Grid item className={styles.bottomIcons}>
        <AddServer />
        <SearchServer />
      </Grid>
    </Grid>
  )
})
