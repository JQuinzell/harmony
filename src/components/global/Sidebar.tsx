import { makeStyles } from '@material-ui/core'
import React from 'react'

const servers = [{ name: 'Cool Server' }, { name: 'S2' }, { name: 'S3' }]

const useStyles = makeStyles((theme) => ({
  nav: {
    background: theme.palette.background.paper,
    height: '100%',
    display: 'inline-block',
  },
  server: {
    width: '70px',
    height: '70px',
    margin: `${theme.spacing(2)}px`,
    borderRadius: '100%',
    background: theme.palette.primary.light,
    textAlign: 'center',
    lineHeight: '70px',
    fontSize: 24,
  },
}))

export const SideBar: React.FC = () => {
  const styles = useStyles()

  return (
    <nav className={styles.nav}>
      {servers.map(({ name }) => (
        <div key={name} className={styles.server}>
          {name.slice(0, 2)}
        </div>
      ))}
    </nav>
  )
}
