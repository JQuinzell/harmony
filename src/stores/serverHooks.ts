import { useRootStore } from '../RootStoreContext'
import { CreateServer } from './interfaces'

export function useServerStore() {
  const rootStore = useRootStore()
  return {
    joinedServers: rootStore.joinedServers,
    currentServer: rootStore.currentServer,
    createServer(data: CreateServer) {
      return rootStore.createServer(data)
    },
    selectServer(title: string) {
      return rootStore.selectServer(title)
    },
    postMessage(text: string) {
      return rootStore.postMessage(text)
    },
    login(username: string, password: string) {
      return rootStore.login(username, password)
    },
  }
}
