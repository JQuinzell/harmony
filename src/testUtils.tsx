import { render } from '@testing-library/react'
import { RootStoreProvider } from './RootStoreContext'
import React from 'react'
import { MemoryRouter } from 'react-router'

export function renderWithProviders(element: React.ReactElement) {
  return render(
    <RootStoreProvider>
      <MemoryRouter>{element}</MemoryRouter>
    </RootStoreProvider>
  )
}
