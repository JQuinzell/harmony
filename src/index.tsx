import React from 'react'
import ReactDOM from 'react-dom'
import { RootStoreProvider } from './RootStoreContext'
import { Routes } from './components/Routes'
import { configure } from 'mobx'

configure({
  enforceActions: 'always',
  computedRequiresReaction: true,
  reactionRequiresObservable: true,
  observableRequiresReaction: true,
  disableErrorBoundaries: true,
})

ReactDOM.render(
  <RootStoreProvider>
    <Routes />
  </RootStoreProvider>,
  document.getElementById('app')
)
