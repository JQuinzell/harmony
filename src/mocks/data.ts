import { ServerPreview } from 'src/stores/RootStore'

interface ServerPreviewsQuery {
  user: { servers: ServerPreview[] }
  servers: ServerPreview[]
}

export const serverPreviewsQuery: ServerPreviewsQuery = {
  user: {
    servers: [
      {
        id: 0,
        title: 'My Server',
        description: 'my server description',
        image: 'my image',
      },
    ],
  },
  servers: [
    {
      id: 1,
      title: 'Server 1',
      description: 'server 1 description',
      image: 'image url',
    },
  ],
}

interface LoginQuery {
  result: string
}

export const loginQuery: LoginQuery = {
  result: 'token',
}
