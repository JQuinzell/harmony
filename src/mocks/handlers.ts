import { graphql } from 'msw'
import { loginQuery, serverByNameQuery, serverPreviewsQuery } from './data'

export const handlers = [
  graphql.query('serverPreviews', (req, res, ctx) => {
    return res(ctx.data(serverPreviewsQuery))
  }),

  graphql.query('login', (req, res, ctx) => {
    return res(ctx.data(loginQuery))
  }),

  graphql.query('serverByName', (req, res, ctx) => {
    return res(ctx.data(serverByNameQuery))
  }),
]
