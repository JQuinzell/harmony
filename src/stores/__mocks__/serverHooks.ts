import { serverPreviewsQuery } from '~/mocks/data'

export const useServerStore = jest.fn().mockReturnValue({
  joinedServers: serverPreviewsQuery.user.servers,
  createServer: jest.fn().mockResolvedValue(undefined),
  selectServer: jest.fn().mockResolvedValue(undefined),
})
