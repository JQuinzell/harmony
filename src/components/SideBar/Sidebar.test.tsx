import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import { serverPreviewsQuery } from 'src/mocks/data'
import { renderWithProviders } from 'src/testUtils'
import { SideBar } from './Sidebar'
import { useServerStore } from 'src/stores/serverHooks'

jest.mock('src/stores/serverHooks')

describe('Sidebar', () => {
  const init = () => renderWithProviders(<SideBar />)
  const mockUseServerStore = (useServerStore as unknown) as jest.MockedFunction<
    typeof useServerStore
  >
  const { selectServer } = mockUseServerStore()
  const titles = serverPreviewsQuery.user.servers.map(({ title }) => title)

  it('should list server and action buttons', async () => {
    init()

    screen.getByRole('button', { name: /add/i })
    screen.getByRole('button', { name: /search/i })
    await waitFor(() =>
      titles.map((title) =>
        expect(screen.getByRole('button', { name: title })).toBeInTheDocument()
      )
    )
  })

  it('select server that is clicked', async () => {
    const title = titles[0]
    init()

    const button = await screen.findByRole('button', { name: title })
    userEvent.click(button)

    expect(selectServer).toBeCalledWith(title)
  })
})
