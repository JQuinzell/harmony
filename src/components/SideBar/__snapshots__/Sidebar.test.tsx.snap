// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should list servers: Renders Sidebar 1`] = `
<div>
  <nav
    class="MuiGrid-root makeStyles-nav-1 MuiGrid-container MuiGrid-direction-xs-column"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <div
        data-testid="My Server"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-server-3 MuiAvatar-colorDefault"
        >
          My
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <div
        data-testid="My Server 2"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-server-3 MuiAvatar-colorDefault"
        >
          My
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root makeStyles-bottomIcons-2 MuiGrid-item"
    >
      <div>
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-server-3 MuiAvatar-colorDefault"
        />
      </div>
      <div>
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-server-3 MuiAvatar-colorDefault"
        />
      </div>
    </div>
  </nav>
</div>
`;
